File: RxTango/python/examples/correlate.py
"""Continuously correlated reads of two attributes, reporting their
measured timestamp skew.

Every interval tick, both attributes are read in parallel via
correlate_snapshot (built on rx.zip).  The pair is emitted only when BOTH
reads complete — same guarantee as a bare zip.  If either fails, the tick
is silently dropped.  Correlated.skew is the measured gap between the two
attributes' own DeviceAttribute.time values, not the read-completion time:
"both completed" says nothing on its own about whether the two values
describe the same instant.

Mirrors `TangoTestCorrelate.java`.

Usage:
    python correlate.py [device] [interval-ms]

    device       defaults to tango://localhost:10000/sys/tg_test/1
    interval-ms  defaults to 1000
"""

import asyncio
import sys
from datetime import timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import read_attribute_ts
from rxtango.correlate import correlate_snapshot


async def main() -> None:
    device      = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"
    interval_ms = int(sys.argv[2]) if len(sys.argv) > 2 else 1000

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)

    print(f"Correlating {device}  every {interval_ms} ms  (Ctrl+C to stop)\n")
    print(f"  {'double_scalar':>16}  {'long_scalar':>14}  {'diff':>14}  {'skew (s)':>10}")
    print("  " + "-" * 62)

    # concat_map (not flat_map): each tick's pair is serialized, so printed
    # lines stay in tick order under load.
    rx.interval(timedelta(milliseconds=interval_ms), scheduler=scheduler).pipe(
        ops.concat_map(
            lambda _: correlate_snapshot(
                read_attribute_ts(device, "double_scalar"),
                read_attribute_ts(device, "long_scalar"),
            ).pipe(
                ops.catch(lambda e, _: rx.empty()),  # drop failed tick
            )
        ),
    ).subscribe(
        on_next=lambda c: print(
            f"  {c.values[0]:>+16.6f}  {c.values[1]:>14}  "
            f"{c.values[0] - float(c.values[1]):>+14.6f}  {c.skew:>10.4f}"
        ),
        on_error=lambda e: print(f"  ERROR: {e}", file=sys.stderr),
        scheduler=scheduler,
    )

    await asyncio.Future()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n  stopped.")

File: RxTango/cpp/examples/correlate.cpp
/**
 * Correlated reads — two devices, same polling window; shows diff.
 *
 * Mirrors Python's correlate.py and Java's TangoTestCorrelate.java.
 *
 * Usage:
 *   ./correlate [device1] [attr1] [device2] [attr2] [interval-ms]
 *   defaults: sys/tg_test/1  double_scalar  sys/tg_test/1  double_scalar  500
 */

#include <chrono>
#include <csignal>
#include <iomanip>
#include <iostream>
#include <string>

#include <rxcpp/rx.hpp>
#include <rxtango/rxtango.hpp>

static volatile sig_atomic_t g_running = 1;

int main(int argc, char* argv[]) {
    const std::string dev1  = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";
    const std::string attr1 = argc > 2 ? argv[2] : "double_scalar";
    const std::string dev2  = argc > 3 ? argv[3] : "tango://localhost:10000/sys/tg_test/1";
    const std::string attr2 = argc > 4 ? argv[4] : "double_scalar";
    const int interval_ms   = argc > 5 ? std::stoi(argv[5]) : 500;

    std::signal(SIGINT, [](int) { g_running = 0; });

    std::cout << "Correlating " << dev1 << "/" << attr1
              << " vs " << dev2 << "/" << attr2 << "\n\n"
              << std::setw(16) << attr1 << "  "
              << std::setw(16) << attr2 << "  "
              << std::setw(12) << "diff" << "\n"
              << std::string(48, '-') << "\n";

    auto sub = rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
        .flat_map([dev1, attr1, dev2, attr2](long) {
            return rxcpp::observable<>::zip(
                [](double a, double b) { return std::make_pair(a, b); },
                rxtango::read_attribute<double>(dev1, attr1),
                rxtango::read_attribute<double>(dev2, attr2)
            );
        })
        .subscribe(
            [](std::pair<double,double> p) {
                std::cout << std::setw(16) << p.first << "  "
                          << std::setw(16) << p.second << "  "
                          << std::setw(12) << (p.first - p.second) << "\n";
            },
            [](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
            }
        );

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    sub.unsubscribe();
    return 0;
}
